### PR TITLE
Adding Changes for F1888 Object Expiration Filtering

### PR DESCRIPTION
**Pull request by IBM-diksha**
Regression Result :
[F1888_regression_result_latest.docx](https://github.com/user-attachments/files/17648755/F1888_regression_result_latest.docx)

### CODE REVIEW

File: examples/ibm-cos-bucket/main.tf [hunk 1]
```diff
@@ -643,3 +649,128 @@ resource "ibm_cos_bucket_lifecycle_configuration"  "lifecycle_config" {
...
+				tags{
+					key = "%s"
+					value = "%s"
+				}
+				object_size_greater_than = "%d"
+				object_size_less_than = "%d"
+			}
+    }  
+    rule_id = "id"
+    status = "enable"
```
Comment (avinash1IBM) [inline examples/ibm-cos-bucket/main.tf:738]: indentation issue

File: examples/ibm-cos-bucket/main.tf [hunk 2]
```diff
@@ -643,3 +649,128 @@ resource "ibm_cos_bucket_lifecycle_configuration"  "lifecycle_config" {
...
+				tags{
+					key = "%s"
+					value = "%s"
+				}
+				tags{
+					key = "%s"
+					value = "%s"
+				}
+				
+			}
```
Comment (avinash1IBM) [inline examples/ibm-cos-bucket/main.tf:770]: indentation issue

File: ibm/flex/structures.go [hunk 3]
```diff
@@ -1001,33 +1001,114 @@ func flattenLifecycleRuleFilter(filter *s3.LifecycleRuleFilter) []interface{} {
...
+		m["value"] = aws.StringValue(v)
+	}
+
+	return []interface{}{m}
+}
 
-func flattenAbortIncompleteMultipartUpload(abortIncompleteMultipartUploadInput *s3.AbortIncompleteMultipartUpload) []interface{} {
-	if abortIncompleteMultipartUploadInput == nil {
+func flattenAbortIncompleteMultipartUpload(u *s3.AbortIncompleteMultipartUpload) []interface{} {
+	if u == nil {
```
Comment (avinash1IBM) [inline ibm/flex/structures.go:1088]: why did you change the variable name, the previous one was better right?

File: ibm/flex/structures.go [hunk 4]
```diff
@@ -1001,33 +1001,114 @@ func flattenLifecycleRuleFilter(filter *s3.LifecycleRuleFilter) []interface{} {
...
 }
 
-func LifecylceRuleGet(lifecycleRuleInput []*s3.LifecycleRule) []map[string]interface{} {
-	rules := make([]map[string]interface{}, 0, len(lifecycleRuleInput))
-	if lifecycleRuleInput != nil {
-		for _, lifecyclerule := range lifecycleRuleInput {
+func LifecylceRuleGet(in []*s3.LifecycleRule) []map[string]interface{} {
+	rules := make([]map[string]interface{}, 0, len(in))
+	if in != nil {
+		for _, lifecyclerule := range in {
```
Comment (avinash1IBM) [inline ibm/flex/structures.go:1104]: same here, why was the parameter names changed?

File: website/docs/r/cos_bucket_lifecycle_configuration.markdown [hunk 5]
```diff
@@ -174,13 +174,149 @@ resource "ibm_cos_bucket_lifecycle_configuration"  "lifecycle" {
...
+      and{
+				prefix = "%s"
+				tags{
+					key = "MyObjectTagKey"
+					value = "MyObjectTagValue"
+				}
+				object_size_greater_than = "%d"
+				object_size_less_than = "%d"
+			}
+    }
```
Comment (avinash1IBM) [inline website/docs/r/cos_bucket_lifecycle_configuration.markdown:266]: indentation issue

File: website/docs/r/cos_bucket_lifecycle_configuration.markdown [hunk 6]
```diff
@@ -174,13 +174,149 @@ resource "ibm_cos_bucket_lifecycle_configuration"  "lifecycle" {
...
+      and{
+				tags{
+					key = "MyObjectTagKey1"
+					value = "MyObjectTagValue1"
+				}
+				tags{
+					key = "MyObjectTagKey2"
+					value = "MyObjectTagValue2"
+				}
+
```
Comment (avinash1IBM) [inline website/docs/r/cos_bucket_lifecycle_configuration.markdown:303]: indentation issue

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go [hunk 7]
```diff
@@ -1253,3 +1521,309 @@ func testAccCheckIBMCosBucket_Lifecycle_Configuration_Transition_With_Multiple_R
...
+		bucket_location = ibm_cos_bucket.bucket.region_location
+		lifecycle_rule {
+		   expiration{
+		     days = "%d"
+		   }
+		   filter {
+		     tag{
+				 key = "%s"
+				 value = "%s"
+			 }
```
Comment (avinash1IBM) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go:1596]: indentation issue

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go [hunk 8]
```diff
@@ -1253,3 +1521,309 @@ func testAccCheckIBMCosBucket_Lifecycle_Configuration_Transition_With_Multiple_R
...
+		   filter {
+		     prefix = "%s"
+		   }  
+		   rule_id = "id"
+		   status = "enable"
+	
+		 }
+	  }
+	
+
```
Comment (avinash1IBM) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go:1561]: blank lines

**Review by avinash1IBM — CHANGES_REQUESTED** [review]

File: ibm/flex/structures.go (hunk 3 above)
Comment (IBM-diksha) [reply ibm/flex/structures.go:1088]: I am not sure may be while copying it might have got from the older one.
I have changed it back.

File: examples/ibm-cos-bucket/main.tf [hunk 9]
```diff
@@ -643,3 +649,128 @@ resource "ibm_cos_bucket_lifecycle_configuration"  "lifecycle_config" {
...
+    expiration{
+      days = 1
+    }
+    filter {
+      and{
+        prefix = "%s"
+        tags{
+          key = "%s"
+          value = "%s"
+          }
```
Comment (avinash1IBM) [inline examples/ibm-cos-bucket/main.tf:732]: indentation issue

File: examples/ibm-cos-bucket/main.tf [hunk 10]
```diff
@@ -643,3 +649,128 @@ resource "ibm_cos_bucket_lifecycle_configuration"  "lifecycle_config" {
...
+			  tags{
+          key = "%s"
+          value = "%s"
+			  }
+				
+			}
+    }
```
Comment (avinash1IBM) [inline examples/ibm-cos-bucket/main.tf:771]: indentation issue

File: ibm/flex/structures.go [hunk 11]
```diff
@@ -1001,33 +1001,114 @@ func flattenLifecycleRuleFilter(filter *s3.LifecycleRuleFilter) []interface{} {
...
+
+	if v := andOp.Tags; v != nil {
+		m["tags"] = flattenMultipleTags(v)
+	}
+
+	return []interface{}{m}
+}
+
+func flattenMultipleTags(in []*s3.Tag) []map[string]interface{} {
+	println("Inside lifecycle rule get")
```
Comment (avinash1IBM) [inline ibm/flex/structures.go:1048]: remove this print statement

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go [hunk 12]
```diff
@@ -1253,3 +1521,305 @@ func testAccCheckIBMCosBucket_Lifecycle_Configuration_Transition_With_Multiple_R
...
+		name              = "%s"
+		service           = "cloud-object-storage"
+		plan              = "standard"
+		location          = "global"
+		resource_group_id = data.ibm_resource_group.cos_group.id
+	}
+	resource "ibm_cos_bucket" "bucket" {
+		bucket_name           = "%s"
+		resource_instance_id  = ibm_resource_instance.instance.id
+	    region_location = "%s"
```
Comment (avinash1IBM) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go:1620]: indentation issue

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go [hunk 13]
```diff
@@ -1253,3 +1521,305 @@ func testAccCheckIBMCosBucket_Lifecycle_Configuration_Transition_With_Multiple_R
...
```
Comment (avinash1IBM) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go:1660]: indentation issue

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go [hunk 14]
```diff
@@ -1253,3 +1521,305 @@ func testAccCheckIBMCosBucket_Lifecycle_Configuration_Transition_With_Multiple_R
...
```
Comment (avinash1IBM) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go:1700]: intendatation issue

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go [hunk 15]
```diff
@@ -1253,3 +1521,305 @@ func testAccCheckIBMCosBucket_Lifecycle_Configuration_Transition_With_Multiple_R
...
```
Comment (avinash1IBM) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go:1748]: indentation issue

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go [hunk 16]
```diff
@@ -1253,3 +1521,305 @@ func testAccCheckIBMCosBucket_Lifecycle_Configuration_Transition_With_Multiple_R
...
```
Comment (avinash1IBM) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration_test.go:1798]: indentation issue

**Review by avinash1IBM — CHANGES_REQUESTED** [review]

**Comment by IBM-diksha** [comment]
[Latest_Regression_18_Nov_2024.docx](https://github.com/user-attachments/files/17799891/Latest_Regression_18_Nov_2024.docx)

**Review by avinash1IBM — APPROVED** [review]

**Comment by IBM-diksha** [comment]
[TF_F1888_Latest_regression_19_Nov.docx](https://github.com/user-attachments/files/17808738/TF_F1888_Latest_regression_19_Nov.docx)

**Comment by IBM-diksha** [comment]
[F1888_4_Dec_regression.docx](https://github.com/user-attachments/files/18007715/F1888_4_Dec_regression.docx)

File: ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration.go [hunk 17]
```diff
@@ -105,14 +105,80 @@ func ResourceIBMCOSBucketLifecycleConfiguration() *schema.Resource {
...
+													Required: true,
+												},
+												"value": {
+													Type:     schema.TypeString,
+													Required: true,
+												},
+											},
+										},
+									},
+									"and": {
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration.go:143]: Can user provide combination of multiple filters along with other filter (tags, prefix...so on) if not can we add some conflicts with can't provider filter.0.and with other arguments

Comment (IBM-diksha) [reply ibm/service/cos/resource_ibm_cos_bucket_lifecycle_configuration.go:143]: We cannot add other filters with each other.
But the parent element lifecycle_rule can multiple rules so we cant add the conflict like lifecycle.0.filter. to the arguments.

**Comment by hkantare** [comment]
Can you fix conflicts

**Review by hkantare — DISMISSED** [review]

**Review by hkantare — APPROVED** [review]
